lib/puppet/configurer.rb

```diff
@@ -229,6 +229,9 @@ def run_internal(options)
       end
 
       execute_prerun_command or return nil
+
+      options[:report].code_id = catalog.code_id
+      options[:report].catalog_uuid = catalog.catalog_uuid
```

Comment:
Good questions. We actually do want to preserve compatibility with new agents & old masters and vice-versa.

#### New agents & Older masters

Older masters won't send `catalog_uuid` so I would expect the agent to generate a new `catalog_uuid` when it deserializes the `/puppet/v3/catalog` response and include that in the report.

Masters before 4.3 and won't send `code_id`, and the agent will [ignore that it's missing](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/resource/catalog.rb#L383-L385), so the call to `catalog.code_id` should return `nil`.

FOSS 4.3.0 masters will send `code_id` with a `nil` value. And PE masters based on 4.3 will send the git sha. So I'd expect the agent to just transfer whatever is in the catalog to the report.

#### Old agents & New masters

Older agents won't send `code_id` or `catalog_uuid` in the report, and newer masters should ignore that they are missing, similar to how the agent ignores when `catalog_uuid` is missing.

There's a separate ticket to verify network compatibility as I'm making a lot of "it should do this" statements, and it'd be good to check that.